tasty-plutus/src/Test/Tasty/Plutus/Internal/TestScript.hs

```diff
@@ -0,0 +1,288 @@
+ > testMintingPolicy =
+ >  mkTestMintingPolicy
+ >    $$(compile [||myMintingPolicy||])
+ >    $$(compile [||toTestMintingPolicy||])
+
+ @since 6.0
+-}
+mkTestMintingPolicy ::
+  forall (redeemer :: Type) (ctx :: Type).
+  CompiledCode (redeemer -> ctx -> Bool) ->
+  CompiledCode ((redeemer -> ctx -> Bool) -> WrappedMintingPolicy) ->
+  TestScript ( 'ForMinting redeemer)
+mkTestMintingPolicy mp wr =
+  TestMintingPolicy $
+    mkMintingPolicyScript $
+      applyCode $$(compile [||getWrappedMintingPolicy||]) $
+        wr `applyCode` mp
+
+{- | 'mkTestMintingPolicyUnsafe' provide the way for creating 'TestScript' with
+ the given 'MintingPolicy', parameterized with any redeemer type.
+ You can use it in case you going to test something tricky
+ and need to break types compatibility.
+
+  = Usage
+
+ > myMintingPolicy :: SomeRedeemerType -> ScriptContext -> Bool
+ > myMintingPolicy r ctx = ...
+
+ > testMintingPolicy :: TestScript ( 'ForMinting SomeOtherRedeemerType)
+ > testMintingPolicy =
+ >  mkTestMintingPolicyUnsafe
+ >    $$(compile [||myMintingPolicy||])
+ >    $$(compile [||toTestMintingPolicy||])
+
+ @since 6.0
+-}
+mkTestMintingPolicyUnsafe ::
+  forall (r :: Type) (redeemer :: Type) (ctx :: Type).
+  CompiledCode (redeemer -> ctx -> Bool) ->
+  CompiledCode ((redeemer -> ctx -> Bool) -> WrappedMintingPolicy) ->
+  TestScript ( 'ForMinting r)
+mkTestMintingPolicyUnsafe mp wr =
+  TestMintingPolicy $
+    mkMintingPolicyScript $
+      applyCode $$(compile [||getWrappedMintingPolicy||]) $
+        wr `applyCode` mp
+
+{- | A wrapper for validators. Use this to construct 'TestScript ForSpending'
+
+ = Usage
+
+ > testValidator :: TestScript ( 'ForSpending SomeType SomeOtherType)
+ > testValidator =
+ >  mkTestValidator
+ >    $$(compile [||myValidator||])
+ >    $$(compile [||toTestValidator||])
+
+ = Important note
+
+ If @myValidator@ requires \'burned in\' arguments, these should be passed via
+ 'liftCode' and 'applyCode', rather than as literal arguments inside of
+ 'compile':
+
+ > testValidatorWithArg ::
+ >    ArgumentType ->
+ >    TestScript ( 'ForSpending SomeType SomeOtherType)
+ > testValidatorWithArg arg =
+ >  mkTestValidator
+ >    ( $$(compile [||myValidator||])
+ >          `applyCode`
+ >          liftCode arg
+ >    )
+ >    $$(compile [||toTestValidator||])
+
+ @since 3.0
+-}
+{-# INLINEABLE toTestValidator #-}
+toTestValidator ::
+  forall (datum :: Type) (redeemer :: Type) (ctx :: Type).
+  (FromData datum, FromData redeemer, FromData ctx) =>
+  (datum -> redeemer -> ctx -> Bool) ->
+  WrappedValidator
+toTestValidator f = WrappedValidator $ \d r p ->
+  case fromBuiltinData d of
+    Nothing -> reportParseFailed "Datum"
+    Just d' -> case fromBuiltinData r of
+      Nothing -> reportParseFailed "Redeemer"
+      Just r' -> case fromBuiltinData p of
+        Nothing -> reportParseFailed "ScriptContext"
+        Just p' ->
+          if f d' r' p'
+            then reportPass
+            else reportFail
```

Comment:
Could this be nicer with do notation?

Reply:
Do you have any suggestions how to do this?

Reply:
@samuelWilliams99 I don't think it's possible in this case. You're probably thinking of something like

```haskell
toTestValidator f = WrappedValidator $ \d r p -> do
    d' <- fromBuiltinData d
    r' <- fromBuiltinData r
    ...
```

This unfortunately won't work: if we get a `Nothing` at the end of that block, we don't know where said `Nothing` came from: it could have been a bad datum parse, a bad redeemer parse, etc. We need this information in order to ensure that we report the error accurately.

It _might_ be possible to do something involving `MonadError` and `handle` or something, but I honestly think this is overkill.

Reply:
Yeah, i was thinking perhaps some kind of helper function that lifting the Maybe to whatever `reportParseFailed` gave, but seems not

Reply:
Yeah, that's basically what the `MonadError`-and-`handle`-based solution would do, but this is definitely no clearer.